python/pyspark/testing/utils.py

```diff
@@ -222,6 +223,114 @@ def check_error(
         )
 
 
+def assertSchemaEqual(actual: StructType, expected: StructType):
+    """
+    A util function to assert equality between DataFrame schemas `actual` and `expected`.
+
+    .. versionadded:: 3.5.0
+
+    Parameters
+    ----------
+    actual : StructType
+        The DataFrame schema that is being compared or tested.
+    expected : StructType
+        The expected schema, for comparison with the actual schema.
+
+    Examples
+    --------
+    >>> from pyspark.sql.types import StructType, StructField, ArrayType, IntegerType, DoubleType
+    >>> s1 = StructType([StructField("names", ArrayType(DoubleType(), True), True)])
+    >>> s2 = StructType([StructField("names", ArrayType(DoubleType(), True), True)])
+    >>> assertSchemaEqual(s1, s2)
+
+    Pass, schemas are identical
+
+    >>> s1 = StructType([StructField("names", StructType(
+    ...                                         [StructField("age", DoubleType(), True)]), True)])
+    >>> s2 = StructType([StructField("first name", StructType([StructField(
+    ...                                                      "age", IntegerType(), True)]), True)])
+    >>> assertSchemaEqual(s1, s2)  # fail  # doctest: +IGNORE_EXCEPTION_DETAIL
+    Traceback (most recent call last):
+    ...
+    PySparkAssertionError: [DIFFERENT_SCHEMA] Schemas do not match.
+    The diff below overlays `actual` and `expected` schema tree strings.
+    - indicates a line that should be removed from `actual` to match `expected`.
+    + indicates a line that should be added to `actual` to match `expected`.
+      root
+    -  |-- names: struct (nullable = true)
+    ?          -
+
+    +  |-- first name: struct (nullable = true)
+    ?     ++++++
+
+    -  |    |-- age: double (nullable = true)
+    ?                ^^^^^
+
+    +  |    |-- age: integer (nullable = true)
+    ?                ^^^ +++
+    """
+    if not isinstance(actual, StructType):
+        raise PySparkAssertionError(
+            error_class="UNSUPPORTED_DATA_TYPE",
```

Comment:
Let's not use the error class UNSUPPORTED_DATA_TYPE here. We can just throw a pyspark type error here and tell users how to make it right (expect a structtype, but got... )